### PR TITLE
Fix possible file corruption on file open for writing

### LINKED ISSUE #5664: File is erased on save when there is no free space left on disk
### Description of the Issue

Saving file when there is no free space left on disk leads to erasing it and data loss.

### Steps to Reproduce the Issue

1. Open existing text file from completely full disk (no free space left)
2. Add several lines.
3. Save the file.

### Expected Behavior

The expected behavior is to report error while saving (`Save failed` or `No space left on disk`) but to retain the initial file data (just as it was before opening and editing it).

### Actual Behavior

The save fails but the file is emptied so all its data before the edit is lost.

### Root cause
The problem is caused by
`FILE *fp = UnicodeConvertor.fopen(fullpath, TEXT("wb"))`
in `FileManager::saveBuffer`

`fopen` with `wb` mode opens the file for writing but if the file exists ERASES ALL ITS DATA.

On the following `fwrite` failure the erased data is not recovered, the file is closed and left with zero size (data is lost).

### Possible solutions
1) Try saving the file under different name in the same folder. Only when the save is successful erase the original file and rename the new one to the original name.

2) Examine the free space left on disk BEFORE opening the file for writing with `fopen`. If there is not enough space DO NOT open the file and immediately return with error `Not enough space`. `GetDiskFreeSpaceEx` WinAPI function can be used to get the remaining free space.

### Notes on solutions
Solution 1 requires enough free disk space to save the copy of the file and this might be a problem in some cases (saving very big file in almost full disk).
For example if the disk is full, we open file from it for editing, we delete portion of it (actually reducing its size) and try saving it, solution 1 will fail (no space on disk for new copy) but solution 2 might be implemented to work in that case.

### PR DESCRIPTION
**Pull request by pnedev**
Microsoft documentation for reference:
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=vs-2019
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fseek-fseeki64?view=vs-2019

Fix #5664

### CODE REVIEW

**Comment by pnedev** [comment]
@donho ,
This PR is based on https://github.com/notepad-plus-plus/notepad-plus-plus/pull/8992 because I believe it is the right approach to deal with this problem but I tested it with Notepad++ 7.9 32bit on Win10 (64bit) and on my setup behaves exactly as supposed to.

`_chsize_s` is called ONLY when it needs to be called (when the file is shortened/shrunk).

File: PowerEditor/src/Utf8_16.cpp [hunk 1]
```diff
@@ -442,6 +457,15 @@ void Utf8_16_Write::fclose()
 
 	if (m_pFile)
 	{
+		if (!keepOriginalLength && !m_bWriteFailure)
```
Comment (donho) [inline PowerEditor/src/Utf8_16.cpp:460]: > _chsize_s is called ONLY when it needs to be called (when the file is shortened/shrunk).

No, it's not what I want. As what I said, _chsize_s should be called only after writing on  disk failure.
This condition let all normal cases have this special treatment.

**Comment by donho** [comment]
Also I tested this PR, it didn't fix the bug - the same result as the current one (before applying this PR).

File: PowerEditor/src/Utf8_16.cpp (hunk 1 above)
Comment (pnedev) [reply PowerEditor/src/Utf8_16.cpp:460]: @donho ,

Calling `_chsize_s ` after the write failure is pointless (or very bad hack to say the least if your idea is to extend the file to its initial length after the fail).

**Comment by pnedev** [comment]
@donho ,
 
> Also I tested this PR, it didn't fix the bug - the same result as the current one (before applying this PR).

Just curious how did you test it?
The PR perfectly works on my setup.

**Comment by donho** [comment]
@pnedev 
1. I use VeraCrypt to create a small drive:

image


2. I added 3 larges JPG to make disk almost full, then I copied change.log, opened it in Notepad++, modified it then save it:

image


3. Then I copied the content, pasted it repeatedly and crazily: 

image

4. I saved it, then the following dialog appeared:

image


5. Answered No, another dialog came:

image


6. Answer No again. Reload the file from disk, here the result:

image

**Comment by pnedev** [comment]
Thanks for the detailed description @donho .

I suppose it is not very "clean" test or one that can be fully trusted because we don't know for sure how VeraCrypt handles it's virtual drives.
Do you know what is the file system of the VeraCrypt drive?

Anyway, no matter what we discuss here you've made-up your mind that this PR's approach is dangerous, not normal or whatever. So I'll test one last thing and if it works I'll make a last PR trying to minimize the impact of this serious bug.

**Comment by donho** [comment]
> I suppose it is not very "clean" test or one that can be fully trusted because we don't know for sure how VeraCrypt handles it's virtual drives.

It's the only way I found, sorry for not having a 3 MB USB flash drive in 2020.
OTOH, I don't see the difference once the drive is mounted by the system.

> Anyway, no matter what we discuss here you've made-up your mind that this PR's approach is dangerous, not normal or whatever.

As I said, and I say again: There's no reason to change current working code for fixing this bug. So let's treat only the special case.

**Pull request closed without merging**